src/storage/src/hummock/local_version/flush_controller.rs

```diff
@@ -0,0 +1,411 @@
+use tokio::sync::{mpsc, oneshot};
+use tracing::{error, info};
+
+use crate::hummock::local_version::local_version_manager::LocalVersionManager;
+use crate::hummock::local_version::upload_handle_manager::UploadHandleManager;
+use crate::hummock::local_version::SyncUncommittedDataStage;
+use crate::hummock::shared_buffer::{SharedBufferEvent, WriteRequest};
+use crate::hummock::{HummockError, HummockResult, SstableIdManagerRef, TrackerId};
+use crate::store::SyncResult;
+
+pub(crate) struct BufferTracker {
+    flush_threshold: usize,
+    block_write_threshold: usize,
+    global_buffer_size: Arc<AtomicUsize>,
+    pub(crate) global_upload_task_size: Arc<AtomicUsize>,
+
+    pub(crate) buffer_event_sender: mpsc::UnboundedSender<SharedBufferEvent>,
+}
+
+impl BufferTracker {
+    pub fn new(
+        flush_threshold: usize,
+        block_write_threshold: usize,
+        buffer_event_sender: mpsc::UnboundedSender<SharedBufferEvent>,
+    ) -> Self {
+        assert!(
+            flush_threshold <= block_write_threshold,
+            "flush threshold {} is not less than block write threshold {}",
+            flush_threshold,
+            block_write_threshold
+        );
+        info!(
+            "buffer tracker init: flush threshold {}, block write threshold {}",
+            flush_threshold, block_write_threshold
+        );
+        Self {
+            flush_threshold,
+            block_write_threshold,
+            global_buffer_size: Arc::new(AtomicUsize::new(0)),
+            global_upload_task_size: Arc::new(AtomicUsize::new(0)),
+            buffer_event_sender,
+        }
+    }
+
+    pub fn get_buffer_size(&self) -> usize {
+        self.global_buffer_size.load(Relaxed)
+    }
+
+    pub fn get_upload_task_size(&self) -> usize {
+        self.global_upload_task_size.load(Relaxed)
+    }
+
+    pub fn can_write(&self) -> bool {
+        self.get_buffer_size() <= self.block_write_threshold
+    }
+
+    pub fn try_write(&self, size: usize) -> bool {
+        loop {
+            let current_size = self.global_buffer_size.load(Acquire);
+            if current_size > self.block_write_threshold {
+                return false;
+            }
+            if self
+                .global_buffer_size
+                .compare_exchange(current_size, current_size + size, Acquire, Acquire)
+                .is_ok()
+            {
+                break true;
+            }
+        }
+    }
+
+    /// Return true when the buffer size minus current upload task size is still greater than the
+    /// flush threshold.
+    pub fn need_more_flush(&self) -> bool {
+        self.get_buffer_size() > self.flush_threshold + self.get_upload_task_size()
+    }
+
+    pub fn send_event(&self, event: SharedBufferEvent) {
+        self.buffer_event_sender.send(event).unwrap();
+    }
+}
+
+pub(crate) struct FlushController {
+    local_version_manager: Arc<LocalVersionManager>,
+    buffer_tracker: Arc<BufferTracker>,
```

Comment:
Do we need to wrap `BufferTacker` in Arc? If we wrap it in Arc, we don't need to wrap the two atomics in Arc.

Reply:
Seems unnecessary. I have removed the `Arc` around `BufferTracker`